cassandra/cluster.py

```diff
@@ -3495,22 +3495,30 @@ def _cancel_timer(self):
         if self._timer:
             self._timer.cancel()
 
-    def _on_timeout(self):
-
-        try:
-            self._connection._requests.pop(self._req_id)
-        # This prevents the race condition of the
-        # event loop thread just receiving the waited message
-        # If it arrives after this, it will be ignored
-        except KeyError:
+    def _on_timeout(self, attempts=0):
+        # PYTHON-853: for short timeouts, we sometimes race with our __init__
+        if self._connection is None and attempts < 3:
+            self._timer = self.session.cluster.connection_class.create_timer(
+                0.01,
+                partial(self._on_timeout, attempts=attempts + 1)
+            )
```

Comment:
Since the looping issue is only valid for a particular case, I would tend to leave the main API untouched. Especially because the `attempts` parameter is only taken into account when connection is None... which is not clear with the function signature. What about something hidden like:
```python
if self._connection is None:
   timer_threshold = getattr(self, '_timer_threshold', 0)
   self._timer_threshold = timer_threshold + 1
   if self._timer_threshold < 3:
       self._timer = self.session.cluster.connection_class.create_timer(
            0.01,
            partial(self._on_timeout, attempts=attempts + 1)
       )
   return
```

Reply:
That'd work just as well, though I prefer using the kwarg. I think the parameter makes it clear that the counter is only ever passed in. With the attribute soluion, it's not obvious that it's never set in some other method.

I'm not too worried about adding to the `_on_timeout` API since it's internal and since we can provide a default.

If I add a docstring/comment explaining `attempts`, would that address your concerns? We also could rename it to `_attempts` to make it clear that you should only use it in rare cases.

Reply:
Can't say I agree with this, but that's fine. I just feel that attempts doesn't make sense to be an API parameter, even if it's internal. e.g. From an API perspective, what does it mean? I'm OK with merging anyway... at least we definitively avoid the rare potential looping issue with this. :+1:

Reply:
My thought was that this is a psuedo-recursive function now, so using a parameter would be a simple way to track depth. Thanks for the feedback, sorry to override it here -- I'll add a comment and merge when the tests look good.